monero-sys/src/lib.rs

```diff
@@ -0,0 +1,1511 @@
+                .createWallet(&path, &password, &language, network_type, kdf_rounds);
+
+        if wallet_pointer.is_null() {
+            anyhow::bail!("Failed to create wallet, got null pointer");
+        }
+
+        let raw_wallet = RawWallet::new(wallet_pointer);
+        let wallet = FfiWallet::new(raw_wallet, background_sync, daemon)
+            .context(format!("Failed to initialize wallet `{}`", &path))?;
+
+        Ok(wallet)
+    }
+
+    /// Create a new wallet from keys or open if it already exists.
+    #[allow(clippy::too_many_arguments)]
+    pub fn open_or_create_wallet_from_keys(
+        &mut self,
+        path: &str,
+        password: Option<&str>,
+        network: monero::Network,
+        address: &monero::Address,
+        view_key: monero::PrivateKey,
+        spend_key: monero::PrivateKey,
+        restore_height: u64,
+        background_sync: bool,
+        daemon: Daemon,
+    ) -> Result<FfiWallet> {
+        if self.wallet_exists(path) {
+            tracing::info!(wallet=%path, "Wallet already exists, opening it");
+
+            return self
+                .open_wallet(path, password, network, background_sync, daemon.clone())
+                .context(format!("Failed to open wallet `{}`", &path));
+        }
+
+        let pathbuf = PathBuf::from(path);
+        if let Some(directory) = pathbuf.parent() {
+            tracing::debug!(
+                "Making sure to create wallet directory `{}`",
+                directory.display()
+            );
+            std::fs::create_dir_all(directory).context(format!(
+                "failed to create wallet directory `{}`",
+                directory.display()
+            ))?;
+        }
+
+        let path = pathbuf.display().to_string();
+
+        tracing::debug!(restore_height, %address, "Creating wallet from keys");
+
+        let_cxx_string!(path = path);
+        let_cxx_string!(password = password.unwrap_or(""));
+        let_cxx_string!(language = "English");
+        let network_type = network.into();
+        let_cxx_string!(address = address.to_string());
+        let_cxx_string!(view_key = view_key.to_string());
+        let_cxx_string!(spend_key = spend_key.to_string());
+        let kdf_rounds = Self::DEFAULT_KDF_ROUNDS;
+
+        let wallet_pointer = self.inner.pinned().createWalletFromKeys(
+            &path,
+            &password,
+            &language,
+            network_type,
+            restore_height,
+            &address,
+            &view_key,
+            &spend_key,
+            kdf_rounds,
+        );
+
+        if wallet_pointer.is_null() {
+            anyhow::bail!("Failed to create wallet from keys, got null pointer");
+        }
+
+        let raw_wallet = RawWallet::new(wallet_pointer);
+        tracing::debug!(path=%path, "Created wallet from keys, initializing");
+        let wallet = FfiWallet::new(raw_wallet, background_sync, daemon)
+            .context(format!("Failed to initialize wallet `{}`", &path))?;
```

Comment:
Add context here about how the wallet was created (from keys) to distinguish the error from errors originating from other wallet open functions